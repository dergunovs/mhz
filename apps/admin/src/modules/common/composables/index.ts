import { useRoute } from 'vue-router';
import { useQueryClient } from '@tanstack/vue-query';

const queryClient = useQueryClient(); // стоит проверить, вызвав invalidate

export function useInvalidate(queryKey: string) {
  const route = useRoute(); 
  function invalidate() {
    if (route.query.create === '1') {
      queryClient.invalidateQueries({ queryKey: [queryKey], exact: true });
    }
  }

  return {
    invalidate,
  };
}
