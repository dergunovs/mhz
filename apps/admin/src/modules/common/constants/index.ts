export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
