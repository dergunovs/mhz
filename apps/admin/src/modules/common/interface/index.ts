import { ICategory, IManager, IManufacturer, IProduct } from 'mhz-types';

export interface ISearchResult {
  products: IProduct[];
  categories: ICategory[];
  manufacturers: IManufacturer[];
  managers: IManager[];
}
