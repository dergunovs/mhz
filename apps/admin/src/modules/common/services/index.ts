import { Ref } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import {
  API_UPLOAD,
  API_UPLOAD_SINGLE,
  API_SEARCH,
  API_STATS_COUNT,
  ISearchResults,
  IEntitiesCount,
  IBaseReply,
} from 'mhz-contracts';

export function search(query: Ref<string>, isAdmin?: boolean) {
  async function fn() {
    const { data } = await api.get<ISearchResults>(API_SEARCH, { params: { search: query.value, isAdmin } });

    return data;
  }

  return useQuery({ queryKey: [API_SEARCH, query], queryFn: fn, enabled: false });
}

export function getEntitiesCount() {
  async function fn() {
    const { data } = await api.get<IEntitiesCount>(API_STATS_COUNT);

    return data;
  }

  return useQuery({ queryKey: [API_STATS_COUNT], queryFn: fn });
}

export function uploadFile(options: object, width?: string) {
  async function fn(file?: File) {
    if (!file) throw new Error();

    const formData = new FormData();

    formData.append('file', file);

    const { data } = await api.post<string>(API_UPLOAD_SINGLE, formData, { params: { width } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function uploadFiles(options: object, width?: string, thumb?: boolean) {
  async function fn(files: File[]) {
    const formData = new FormData();

    files.forEach((file) => formData.append('files', file));

    const { data } = await api.post<string[]>(API_UPLOAD, formData, { params: { width, thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function deleteFile(thumb?: boolean) {
  async function fn(filename: string) {
    const { data } = await api.delete<IBaseReply>(`${API_UPLOAD}/${filename}`, { params: { thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn });
}
