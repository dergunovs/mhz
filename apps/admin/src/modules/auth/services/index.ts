import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
