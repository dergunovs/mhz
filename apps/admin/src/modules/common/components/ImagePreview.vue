<template>
  <div :class="$style.container">
    <div>
      Image<template v-if="props.urls.length > 1">s <i>(sortable)</i></template>
    </div>

    <Sortable
      :list="props.urls"
      :itemKey="(item) => item"
      tag="div"
      @end="updateIndex"
      :class="$style.images"
      data-test="image-preview-sortable"
    >
      <template #item="{ element }">
        <div :class="$style.image" :key="element">
          <img :src="`${PATH_UPLOAD}/${element}`" width="200" alt="Image" loading="lazy" crossorigin="anonymous" />
          <UiButton @click="handleDeleteFile(element)" layout="plain" data-test="image-preview-delete">Delete</UiButton>
        </div>
      </template>
    </Sortable>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, watch } from 'vue';

import { Sortable } from 'sortablejs-vue3';

import { UiButton } from 'mhz-ui';

import { PATH_UPLOAD } from '@/common/constants';
import { deleteFile } from '@/common/services';

interface IProps {
  urls: string[];
  isThumb?: boolean;
}

const props = defineProps<IProps>();
const emit = defineEmits(['delete', 'update']);

const urlsSortable = ref<string[]>([]);

function updateIndex(event: { oldIndex: number; newIndex: number }) {
  urlsSortable.value.splice(event.newIndex, 0, urlsSortable.value.splice(event.oldIndex, 1)[0]);
  emit('update', urlsSortable.value);
}

const { mutate: mutateDeleteFile } = deleteFile(props.isThumb);

function handleDeleteFile(url: string) {
  mutateDeleteFile(url);
  emit('delete', url, true);
}

function getUrls() {
  urlsSortable.value = [...props.urls];
}

onMounted(() => {
  getUrls();

  watch(
    () => props.urls,
    () => getUrls()
  );
});
</script>

<style module lang="scss">
.container {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.images {
  display: flex;
  gap: 8px;
}

.image {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 200px;
  padding: 8px;
  border: 1px solid var(--color-gray);
  border-radius: 16px;
}
</style>
