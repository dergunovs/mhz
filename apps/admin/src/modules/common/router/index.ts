import { createRouter, createWebHistory } from 'vue-router';

import { routes } from '@/common/router/routes';
import { getCookieToken } from '@/auth/composables';
import { URL_LOGIN } from '@/auth/constants';

const router = createRouter({
  history: createWebHistory(),
  routes,
  scrollBehavior(to, from, savedPosition) {
    return savedPosition ? savedPosition : { top: 0 };
  },
});

router.beforeEach((to, from, next) => {
  if (to.path !== URL_LOGIN && !getCookieToken()) {
    next({ path: `${URL_LOGIN}?logout=1` });
  } else {
    next();
  }
});

export default router;
