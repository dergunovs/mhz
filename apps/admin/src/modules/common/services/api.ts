import axios from 'axios';

import { logout } from '@/auth/composables';

axios.defaults.baseURL = import.meta.env.VITE_API;

export const api = axios;

export function setAuthHeader(token: string) {
  api.defaults.headers.common['Authorization'] = `Bearer ${token}`;
}

export function deleteAuthHeader() {
  api.defaults.headers.common['Authorization'] = ``;
}

export function handleError(error: unknown) {
  if (axios.isAxiosError(error)) {
    if (error.response?.status === 403) logout();
    alert(error.response?.data.message);
  }
}
