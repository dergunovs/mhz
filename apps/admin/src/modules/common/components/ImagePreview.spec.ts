import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { IBaseReply } from 'mhz-contracts';
import { dataTest } from 'mhz-helpers';

import ImagePreview from './ImagePreview.vue';

import { mockMutationReply, wrapperFactory } from '@/common/test';
import * as commonServices from '@/common/services';

const URLS = ['1.jpg', '2.jpg'];

const spyMutateDelete = vi.fn();

vi.spyOn(commonServices, 'deleteFile').mockImplementation(() => {
  return mockMutationReply<IBaseReply, { url: string; isThumb: boolean }>(spyMutateDelete);
});

const imagePreviewSortable = dataTest('image-preview-sortable');
const imagePreviewDelete = dataTest('image-preview-delete');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(ImagePreview, {
    props: {
      urls: URLS,
      isThumb: false,
    },
  });
});

enableAutoUnmount(afterEach);

describe('ImagePreview', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ImagePreview)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('passes props to child sortable component', async () => {
    expect(wrapper.findComponent<DefineComponent>(imagePreviewSortable).vm.$attrs.list).toEqual(URLS);
  });

  it('updates index by child sortable component emit', async () => {
    const OLD_INDEX = 1;
    const NEW_INDEX = 0;

    wrapper
      .findComponent<DefineComponent>(imagePreviewSortable)
      .vm.$emit('end', { oldIndex: OLD_INDEX, newIndex: NEW_INDEX });

    expect(wrapper.emitted()['update']).toHaveLength(1);
    expect(wrapper.emitted()['update'][0]).toEqual([[URLS[OLD_INDEX], URLS[NEW_INDEX]]]);
  });

  it('deletes image by delete button click', async () => {
    expect(spyMutateDelete).toBeCalledTimes(0);

    await wrapper.findComponent(imagePreviewDelete).trigger('click');

    expect(spyMutateDelete).toBeCalledTimes(1);
    expect(wrapper.emitted()['delete']).toHaveLength(1);
  });
});
