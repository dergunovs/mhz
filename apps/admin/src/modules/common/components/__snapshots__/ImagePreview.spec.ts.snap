// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImagePreview > matches snapshot 1`] = `
"<div class="_container_8b61a5">
  <div> Images <i>(sortable)</i></div>
  <div list="1.jpg,2.jpg" itemkey="(item) => item" tag="div" class="_images_8b61a5" data-test="image-preview-sortable">
    <div class="_image_8b61a5"><img src="http://localhost:5000/upload/item" width="200" alt="Image" crossorigin="anonymous"><button layout="plain" data-test="image-preview-delete">Delete</button></div>
  </div>
</div>"
`;
